base/stats.go

```diff
@@ -0,0 +1,747 @@
+	DocPushCount        *SgwIntStat `json:"doc_push_count"`
+	ProposeChangeCount  *SgwIntStat `json:"propose_change_count"`
+	ProposeChangeTime   *SgwIntStat `json:"propose_change_time"`
+	SyncFunctionCount   *SgwIntStat `json:"sync_function_count"`
+	SyncFunctionTime    *SgwIntStat `json:"sync_function_time"`
+	WriteProcessingTime *SgwIntStat `json:"write_processing_time"`
+}
+
+type DatabaseStats struct {
+	AbandonedSeqs           *SgwIntStat `json:"abandoned_seqs"`
+	ConflictWriteCount      *SgwIntStat `json:"conflict_write_count"`
+	Crc32MatchCount         *SgwIntStat `json:"crc32c_match_count"`
+	DCPCachingCount         *SgwIntStat `json:"dcp_caching_count"`
+	DCPCachingTime          *SgwIntStat `json:"dcp_caching_time"`
+	DCPReceivedCount        *SgwIntStat `json:"dcp_received_count"`
+	DCPReceivedTime         *SgwIntStat `json:"dcp_received_time"`
+	DocReadsBytesBlip       *SgwIntStat `json:"doc_reads_bytes_blip"`
+	DocWritesBytes          *SgwIntStat `json:"doc_writes_bytes"`
+	DocWritesBytesBlip      *SgwIntStat `json:"doc_writes_bytes_blip"`
+	DocWritesXattrBytes     *SgwIntStat `json:"doc_writes_xattr_bytes"`
+	HighSeqFeed             *SgwIntStat `json:"high_seq_feed"`
+	NumDocReadsBlip         *SgwIntStat `json:"num_doc_reads_blip"`
+	NumDocReadsRest         *SgwIntStat `json:"num_doc_reads_rest"`
+	NumDocWrites            *SgwIntStat `json:"num_doc_writes"`
+	NumReplicationsActive   *SgwIntStat `json:"num_replications_active"`
+	NumReplicationsTotal    *SgwIntStat `json:"num_replications_total"`
+	NumTombstonesCompacted  *SgwIntStat `json:"num_tombstones_compacted"`
+	SequenceAssignedCount   *SgwIntStat `json:"sequence_assigned_count"`
+	SequenceGetCount        *SgwIntStat `json:"sequence_get_count"`
+	SequenceIncrCount       *SgwIntStat `json:"sequence_incr_count"`
+	SequenceReleasedCount   *SgwIntStat `json:"sequence_released_count"`
+	SequenceReservedCount   *SgwIntStat `json:"sequence_reserved_count"`
+	WarnChannelsPerDocCount *SgwIntStat `json:"warn_channels_per_doc_count"`
+	WarnGrantsPerDocCount   *SgwIntStat `json:"warn_grants_per_doc_count"`
+	WarnXattrSizeCount      *SgwIntStat `json:"warn_xattr_size_count"`
+
+	// These can be cleaned up in future versions of SGW, implemented as maps to reduce amount of potential risk
+	// prior to Hydrogen release.
+	CacheFeedMapStats  *ExpVarMapWrapper `json:"cache_feed"`
+	ImportFeedMapStats *ExpVarMapWrapper `json:"import_feed"`
+}
+
+type ExpVarMapWrapper struct {
+	*expvar.Map
+}
+
+func (wrapper *ExpVarMapWrapper) MarshalJSON() ([]byte, error) {
+	return []byte(wrapper.String()), nil
+}
+
+type ReplicatorStats struct {
+	checkedSentDesc,
+	numAttachmentBytesTransferred,
+	numAttachmentsTransferred,
+	numDocsFailedToPush,
+	numDocsPushed *prometheus.Desc
+
+	toplevelMap *expvar.Map
+}
+
+func (rs *ReplicatorStats) MarshalJSON() ([]byte, error) {
+	return []byte(rs.toplevelMap.String()), nil
+}
+
+func (rs *ReplicatorStats) Describe(ch chan<- *prometheus.Desc) {
+	return
```

Comment:
https://godoc.org/github.com/prometheus/client_golang/prometheus#Desc
https://godoc.org/github.com/prometheus/client_golang/prometheus#Collector

From these two above essentially it looks like Describe has protection against having duplicate descriptions which isn't allowed as obviously this would cause problems when settings a single stat to the same value. A description essentially is the "key" for the stat.

There is no functionality lost by not having it.

In our implementation using this wouldn't work as we do technically have the same description used twice but when we use it we set different label values. The documentation around Collector seems to suggest that this is fine and we don't have to use it provided we can make guarantees ourselves that there won't be conflicts.